`odo dev` freezes then panics when run from `~`

/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:**

RHEL 8.7

**Output of `odo version`:**

```
odo v3.9.0 (8cb53c2a0)

Server: https://api.crc.testing:6443
Kubernetes: v1.24.0+3882f8f
```

## How did you run odo exactly?

(assuming there is no devfile in `~`):

```
cd
odo dev
```

## Actual behavior

It freezes:

![odo-dev-homedir](https://user-images.githubusercontent.com/22376627/234939783-4f835749-4461-4453-a8d4-83bb64e1885f.gif)

Eventually, it panics with the logs below.

## Expected behavior

The interactive wizard to select a devfile starts.

## Any logs, error output, etc?

```
panic: regexp: Compile(`(let|const|var)\s+.profile/* %v *//convert/etc/ssh/gid_map/healthz/metrics/mutate-/uid_map/vendor//version000000000123456715:04:051606nict1694acad1959acad2.5.4.102.5.4.112.5.4.172006-1-238;5;%d;3des-cbc4882812548;5;%d;:*(\d+\s*=\s*([^;]*)`): error parsing regexp: missing closing ): `(let|const|var)\s+.profile/* %v *//convert/etc/ssh/gid_map/healthz/metrics/mutate-/uid_map/vendor//version000000000123456715:04:051606nict1694acad1959acad2.5.4.102.5.4.112.5.4.172006-1-238;5;%d;3des-cbc4882812548;5;%d;:*(\d+\s*=\s*([^;]*)`

goroutine 1 [running]:
regexp.MustCompile({0xc001ff8000, 0xee})
        /usr/lib/golang/src/regexp/regexp.go:319 +0xbb
github.com/redhat-developer/alizer/go/pkg/apis/enricher/framework/javascript/nodejs.getPort({0xc038ce2000, 0x48f81f8}, {0xc0160bc150, 0xffffffffffffffff?, 0x33f38f8?})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/enricher/framework/javascript/nodejs/express_detector.go:86 +0x199
github.com/redhat-developer/alizer/go/pkg/apis/enricher/framework/javascript/nodejs.ExpressDetector.DoPortsDetection({}, 0xc008546960, 0xd?)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/enricher/framework/javascript/nodejs/express_detector.go:51 +0x20b
github.com/redhat-developer/alizer/go/pkg/apis/enricher.JavaScriptEnricher.DoEnrichComponent({}, 0xc008546960, {{0xc00005e0dc, 0xe}, {0xc0008e48e8, 0x3, 0x3}}, 0x4f49c50?)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/enricher/javascript_enricher.go:94 +0x5e8
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.enrichComponent(0xc008546960, {{0xc00005e0dc, 0xe}, {0xc0008e48e8, 0x3, 0x3}}, 0xc0008322a0?)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:331 +0x89
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.detectComponentByAnalyzingConfigFile({0xc00a2673e0, 0x5a}, {0xc000cfff50, 0xa}, {{0xc00005e0dc, 0xe}, {0xc0008e48e8, 0x3, 0x3}}, 0xc001010a90)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:303 +0x3ae
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.detectComponentUsingConfigFile({0xc00a2673e0, 0x5a}, {0xc0010aa040?, 0x2, 0x2}, {{0xc00005e0dc, 0xe}, {0xc0008e48e8, 0x3, 0x3}}, ...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:316 +0x15a
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.DetectComponentsFromFilesList({0xc0193fa000, 0x161375, 0xc001010a90?}, {{0xc00005e0dc, 0xe}, {0xc0008e48e8, 0x3, 0x3}}, 0xc0008e48e8?)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:229 +0x327
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.detectComponentsWithSettings({{0xc00005e0dc, 0xe}, {0xc0008e48e8, 0x3, 0x3}}, 0xc000a34a80?)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:86 +0xa5
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.detectComponentsWithPathAndPortStartegy(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:55
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.DetectComponents({0xc00005e0dc, 0xe})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:34 +0xd1
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.selectDevFilesFromComponentsDetectedInPath({0xc00005e0dc, 0xe}, {0xc000a11200, 0x19, 0x20})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/devfile_recognizer.go:51 +0x65
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.SelectDevFilesFromTypes({0xc00005e0dc, 0xe}, {0xc000a11200, 0x19, 0x20})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/devfile_recognizer.go:28 +0x54
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.SelectDevFileFromTypes(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/devfile_recognizer.go:67
github.com/redhat-developer/odo/pkg/alizer.(*Alizer).DetectFramework(0xc00006c000?, {0x3854d98?, 0xc000b1e000?}, {0xc00005e0dc, 0xe})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/alizer/alizer.go:46 +0x23a
github.com/redhat-developer/odo/pkg/init/backend.(*AlizerBackend).SelectDevfile(0xc000690b00, {0x3854d98?, 0xc000b1e000?}, 0x11b0fc7?, {0xc0007a10f0?, 0xc0007a1150?}, {0xc00005e0dc, 0xe})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/init/backend/alizer.go:38 +0xb0
github.com/redhat-developer/odo/pkg/init.(*InitClient).SelectDevfile(0xc00072c4b0, {0x3854d98, 0xc000b1e000}, 0xc000b1e150, {0x386a6c0, 0x4f49c50}, {0xc00005e0dc, 0xe})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/init/init.go:94 +0x117
github.com/redhat-developer/odo/pkg/init.(*InitClient).SelectAndPersonalizeDevfile(0xc00072c4b0, {0x3854d98, 0xc000b1e000}, 0x21?, {0xc00005e0dc, 0xe})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/init/init.go:277 +0x97
github.com/redhat-developer/odo/pkg/init.(*InitClient).InitDevfile(0xc00072c4b0, {0x3854d98, 0xc000b1e000}, 0xc000b1e090, {0xc00005e0dc, 0xe}, 0xc000b1e0c0, 0xc000690b80)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/init/init.go:323 +0x128
github.com/redhat-developer/odo/pkg/odo/genericclioptions.runPreInit({0x3854d98, 0xc000b1e000}, {0xc00005e0dc, 0xe}, 0xc0005ae000, {0x3865160, 0xc0005987b8}, {0x348db8e, 0x59})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/genericclioptions/init.go:32 +0x1d9
github.com/redhat-developer/odo/pkg/odo/genericclioptions.GenericRun({0x38559d8, 0xc00019a5a0}, 0xc00016a300, {0x4f49c50, 0x0, 0x0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/genericclioptions/runnable.go:242 +0x1334
github.com/redhat-developer/odo/pkg/odo/cli/dev.NewCmdDev.func1(0xc00016a300?, {0x4f49c50?, 0x0?, 0x0?})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/dev/dev.go:235 +0x38
github.com/spf13/cobra.(*Command).execute(0xc00016a300, {0x4f49c50, 0x0, 0x0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc000120000)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:968
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:961
main.main()
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/cmd/odo/odo.go:83 +0x585
```

--- a/docs/website/docs/command-reference/docs-mdx/init/interactive_mode_directory_with_sources_output.mdx
+++ b/docs/website/docs/command-reference/docs-mdx/init/interactive_mode_directory_with_sources_output.mdx
@@ -7,6 +7,7 @@ $ odo init
  \__/
 
 Interactive mode enabled, please answer the following questions:
+ ✓  Determining a Devfile for the current directory [1s]
 Based on the files in the current directory odo detected
 Language: JavaScript
 Project type: Node.js

--- a/docs/website/docs/user-guides/quickstart/docs-mdx/dotnet/dotnet_odo_init_output.mdx
+++ b/docs/website/docs/user-guides/quickstart/docs-mdx/dotnet/dotnet_odo_init_output.mdx
@@ -7,6 +7,7 @@ $ odo init
  \__/
 
 Interactive mode enabled, please answer the following questions:
+ ✓  Determining a Devfile for the current directory [1s]
 Based on the files in the current directory odo detected
 Language: .NET
 Project type: dotnet

--- a/docs/website/docs/user-guides/quickstart/docs-mdx/go/go_odo_init_output.mdx
+++ b/docs/website/docs/user-guides/quickstart/docs-mdx/go/go_odo_init_output.mdx
@@ -7,6 +7,7 @@ $ odo init
  \__/
 
 Interactive mode enabled, please answer the following questions:
+ ✓  Determining a Devfile for the current directory [1s]
 Based on the files in the current directory odo detected
 Language: Go
 Project type: Go

--- a/docs/website/docs/user-guides/quickstart/docs-mdx/java/java_odo_init_output.mdx
+++ b/docs/website/docs/user-guides/quickstart/docs-mdx/java/java_odo_init_output.mdx
@@ -7,6 +7,7 @@ $ odo init
  \__/
 
 Interactive mode enabled, please answer the following questions:
+ ✓  Determining a Devfile for the current directory [1s]
 Based on the files in the current directory odo detected
 Language: Java
 Project type: springboot

--- a/docs/website/docs/user-guides/quickstart/docs-mdx/nodejs/nodejs_odo_init_output.mdx
+++ b/docs/website/docs/user-guides/quickstart/docs-mdx/nodejs/nodejs_odo_init_output.mdx
@@ -7,6 +7,7 @@ $ odo init
  \__/
 
 Interactive mode enabled, please answer the following questions:
+ ✓  Determining a Devfile for the current directory [1s]
 Based on the files in the current directory odo detected
 Language: JavaScript
 Project type: Node.js

--- a/pkg/init/backend/alizer.go
+++ b/pkg/init/backend/alizer.go
@@ -34,39 +34,61 @@ func (o *AlizerBackend) Validate(flags map[string]string, fs filesystem.Filesyst
 	return nil
 }
 
-// SelectDevfile calls thz Alizer to detect the devfile and asks for confirmation to the user
-func (o *AlizerBackend) SelectDevfile(ctx context.Context, flags map[string]string, fs filesystem.Filesystem, dir string) (location *api.DetectionResult, err error) {
-	spinner := log.Spinnerf("Determining a Devfile for the current directory")
-	defer spinner.End(err == nil)
-	selected, defaultVersion, registry, err := o.alizerClient.DetectFramework(ctx, dir)
-	if err != nil {
-		return nil, err
+// SelectDevfile calls the Alizer to detect the devfile and asks for confirmation to the user
+func (o *AlizerBackend) SelectDevfile(ctx context.Context, flags map[string]string, fs filesystem.Filesystem, dir string) (*api.DetectionResult, error) {
+	type result struct {
+		location *api.DetectionResult
+		err      error
 	}
+	var (
+		resultChan = make(chan result)
+	)
+	go func() {
+		location, err := func() (location *api.DetectionResult, err error) {
+			spinner := log.Spinnerf("Determining a Devfile for the current directory")
+			defer spinner.End(err == nil)
+			selected, defaultVersion, registry, err := o.alizerClient.DetectFramework(ctx, dir)
+			if err != nil {
+				return nil, err
+			}
 
-	msg := fmt.Sprintf("Based on the files in the current directory odo detected\nLanguage: %s\nProject type: %s", selected.Language, selected.ProjectType)
+			msg := fmt.Sprintf("Based on the files in the current directory odo detected\nLanguage: %s\nProject type: %s", selected.Language, selected.ProjectType)
 
-	appPorts, err := o.alizerClient.DetectPorts(dir)
-	if err != nil {
-		return nil, err
-	}
-	appPortsAsString := make([]string, 0, len(appPorts))
-	for _, p := range appPorts {
-		appPortsAsString = append(appPortsAsString, strconv.Itoa(p))
-	}
-	if len(appPorts) > 0 {
-		msg += fmt.Sprintf("\nApplication ports: %s", strings.Join(appPortsAsString, ", "))
-	}
+			appPorts, err := o.alizerClient.DetectPorts(dir)
+			if err != nil {
+				return nil, err
+			}
+			appPortsAsString := make([]string, 0, len(appPorts))
+			for _, p := range appPorts {
+				appPortsAsString = append(appPortsAsString, strconv.Itoa(p))
+			}
+			if len(appPorts) > 0 {
+				msg += fmt.Sprintf("\nApplication ports: %s", strings.Join(appPortsAsString, ", "))
+			}
 
-	fmt.Println(msg)
-	fmt.Printf("The devfile \"%s:%s\" from the registry %q will be downloaded.\n", selected.Name, defaultVersion, registry.Name)
-	confirm, err := o.askerClient.AskCorrect()
-	if err != nil {
-		return nil, err
+			fmt.Println(msg)
+			fmt.Printf("The devfile \"%s:%s\" from the registry %q will be downloaded.\n", selected.Name, defaultVersion, registry.Name)
+			confirm, err := o.askerClient.AskCorrect()
+			if err != nil {
+				return nil, err
+			}
+			if !confirm {
+				return nil, nil
+			}
+			return alizer.NewDetectionResult(selected, registry, appPorts, defaultVersion, ""), nil
+		}()
+		resultChan <- result{
+			location: location,
+			err:      err,
+		}
+	}()
+
+	select {
+	case res := <-resultChan:
+		return res.location, res.err
+	case <-ctx.Done():
+		return nil, fmt.Errorf("interrupted: %w", ctx.Err())
 	}
-	if !confirm {
-		return nil, nil
-	}
-	return alizer.NewDetectionResult(selected, registry, appPorts, defaultVersion, ""), nil
 }
 
 func (o *AlizerBackend) SelectStarterProject(devfile parser.DevfileObj, flags map[string]string) (starter *v1alpha2.StarterProject, err error) {

--- a/pkg/init/backend/alizer.go
+++ b/pkg/init/backend/alizer.go
@@ -3,6 +3,7 @@ package backend
 import (
 	"context"
 	"fmt"
+	"github.com/redhat-developer/odo/pkg/log"
 	"strconv"
 	"strings"
 
@@ -35,6 +36,8 @@ func (o *AlizerBackend) Validate(flags map[string]string, fs filesystem.Filesyst
 
 // SelectDevfile calls thz Alizer to detect the devfile and asks for confirmation to the user
 func (o *AlizerBackend) SelectDevfile(ctx context.Context, flags map[string]string, fs filesystem.Filesystem, dir string) (location *api.DetectionResult, err error) {
+	spinner := log.Spinnerf("Determining a Devfile for the current directory")
+	defer spinner.End(err == nil)
 	selected, defaultVersion, registry, err := o.alizerClient.DetectFramework(ctx, dir)
 	if err != nil {
 		return nil, err

--- a/pkg/init/backend/alizer.go
+++ b/pkg/init/backend/alizer.go
@@ -58,6 +58,7 @@ func (o *AlizerBackend) SelectDevfile(ctx context.Context, flags map[string]stri
 			if err != nil {
 				return nil, err
 			}
+			spinner.End(true)
 			appPortsAsString := make([]string, 0, len(appPorts))
 			for _, p := range appPorts {
 				appPortsAsString = append(appPortsAsString, strconv.Itoa(p))

--- a/pkg/init/init.go
+++ b/pkg/init/init.go
@@ -96,6 +96,9 @@ func (o *InitClient) SelectDevfile(ctx context.Context, flags map[string]string,
 		if backend == o.alizerBackend {
 			// Fallback to the Interactive Mode if Alizer could not determine the Devfile.
 			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					return nil, err
+				}
 				log.Warningf("Could not determine a Devfile based on the files in the current directory: %v", err)
 			}
 			return o.interactiveBackend.SelectDevfile(ctx, flags, fs, dir)

--- a/pkg/init/init.go
+++ b/pkg/init/init.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/AlecAivazis/survey/v2/terminal"
 	"net/url"
 	"path/filepath"
 	"strings"
@@ -96,7 +97,7 @@ func (o *InitClient) SelectDevfile(ctx context.Context, flags map[string]string,
 		if backend == o.alizerBackend {
 			// Fallback to the Interactive Mode if Alizer could not determine the Devfile.
 			if err != nil {
-				if errors.Is(err, context.Canceled) {
+				if errors.Is(err, context.Canceled) || errors.Is(err, terminal.InterruptErr) {
 					return nil, err
 				}
 				log.Warningf("Could not determine a Devfile based on the files in the current directory: %v", err)

--- a/pkg/odo/cli/dev/dev.go
+++ b/pkg/odo/cli/dev/dev.go
@@ -264,6 +264,10 @@ func (o *DevOptions) Run(ctx context.Context) (err error) {
 }
 
 func (o *DevOptions) HandleSignal() error {
+	if o.cancel == nil {
+		klog.V(1).Infof("Returning since the cancel is nil; it must be user interrupt while running preInit.")
+		return nil
+	}
 	o.cancel()
 	// At this point, `ctx.Done()` will be raised, and the cleanup will be done
 	// wait for the cleanup to finish and let the main thread finish instead of signal handler go routine from runnable

--- a/pkg/odo/cli/dev/dev.go
+++ b/pkg/odo/cli/dev/dev.go
@@ -263,12 +263,8 @@ func (o *DevOptions) Run(ctx context.Context) (err error) {
 	)
 }
 
-func (o *DevOptions) HandleSignal() error {
-	if o.cancel == nil {
-		klog.V(1).Infof("Returning since the cancel is nil; it must be user interrupt while running preInit.")
-		return nil
-	}
-	o.cancel()
+func (o *DevOptions) HandleSignal(ctx context.Context, cancelFunc context.CancelFunc) error {
+	cancelFunc()
 	// At this point, `ctx.Done()` will be raised, and the cleanup will be done
 	// wait for the cleanup to finish and let the main thread finish instead of signal handler go routine from runnable
 	select {}

--- a/pkg/odo/genericclioptions/runnable.go
+++ b/pkg/odo/genericclioptions/runnable.go
@@ -52,7 +52,7 @@ type Runnable interface {
 }
 
 type SignalHandler interface {
-	HandleSignal() error
+	HandleSignal(ctx context.Context, cancelFunc context.CancelFunc) error
 }
 
 type Cleanuper interface {
@@ -81,9 +81,9 @@ const (
 
 func GenericRun(o Runnable, cmd *cobra.Command, args []string) error {
 	var (
-		err       error
-		startTime = time.Now()
-		ctx       = cmd.Context()
+		err             error
+		startTime       = time.Now()
+		ctx, cancelFunc = context.WithCancel(cmd.Context())
 	)
 
 	defer func() {
@@ -174,7 +174,7 @@ func GenericRun(o Runnable, cmd *cobra.Command, args []string) error {
 	go commonutil.StartSignalWatcher(captureSignals, func(receivedSignal os.Signal) {
 		err = fmt.Errorf("user interrupted the command execution: %w", terminal.InterruptErr)
 		if handler, ok := o.(SignalHandler); ok {
-			err = handler.HandleSignal()
+			err = handler.HandleSignal(ctx, cancelFunc)
 			if err != nil {
 				log.Errorf("error handling interrupt signal : %v", err)
 			}

--- a/pkg/odo/genericclioptions/runnable.go
+++ b/pkg/odo/genericclioptions/runnable.go
@@ -91,6 +91,7 @@ func GenericRun(o Runnable, cmd *cobra.Command, args []string) error {
 			cmd.SilenceErrors = true
 			cmd.SilenceUsage = true
 		}
+		cancelFunc()
 	}()
 
 	userConfig, _ := preference.NewClient(ctx)

